### DISCUSSION TITLE
How to make a glsl shader and or custom node.

### DISCUSSION THREAD

**Discussion by Dai696** [bnpr/Malt#339] [Q&A] [answered]
I'm aware this question has been asked a few times before and you've provided updated answers however I'm still very confused.

I'm trying to write my own sorta shader setup by hopefully overwriting an existing one for learning purposes however I can't seem to find any of the glsl code even with the examples package.

Ideally I'd like to take the pbr example node I read about and impliment a bit of toon shading to it to try and replicate what I'm doing in Unity. For testing I was going to do it to the diffuse shader but I can't even seem to find that.

I also can't seem to get ambient lighting working with the default toon shader, my assumption is that it doesn't read the ambient color by default? I'd like to tint the shadows with it.

![image](https://user-images.githubusercontent.com/17523069/175405208-5183d2a2-f32c-4f82-8b24-88c0437fda42.png)

Lastly I was curious if there were plans to simplify some of the nodes. Note Utils vec2-vec4 being seperate seems really complicated and it'd be nice if there was a way to wrap these into just the utils and have the type adjust automatically (similar to Unity or Unreals shader editors)

![image](https://user-images.githubusercontent.com/17523069/175405405-37e113e5-25e3-40ea-85c0-a505cf302c44.png)

So instead of picking the type, then the util, we can just use the util and the type is done for us. (maybe defaulting to vec4 or something)

We can then also specify the type in the node if needed.

If vec2 -> vec4, it'll add 2 extra 0 inputs to it

If vec4->vec2, it'll ignore the last 2 values.

**Comment by Kolupsy** [comment]
If you want to learn how to make custom nodes for Malt, you can take a look at the plugin that I am currently making. It implements custom nodes and even wrapper nodes (which are optional though) https://github.com/Kolupsy/MaltShadingEssentials
If you just want to make a library of custom nodes, you can ignore almost everything.
What you need is the basic structure of the class in the main `__init__.py` file where you need to override `poll_pipeline` and `register_graph_libraries`.
Other than that you need to just make a directory where you want to define all your shader code ('ShaderFunctions' in my case) and add glsl files. Not that every file in my directory ends with `.internal.glsl` which hides them as custom nodes, in your case you will probably not want to have '.internal.' in the name of the file for now. 

Other than that feel free to use this plugin because it aims to enable an eevee-like noding workflow in Malt. Instructions on how to install and use can be found in the README file.

A very quick way of adding custom nodes though would be to use the library source path on any malt tree to point to a glsl library file
![image](https://user-images.githubusercontent.com/88381426/175412789-ceca7d7e-cbfa-4e95-8904-18cfec852382.png)

**Comment by Dai696** [comment]
Wow this is an amazing plugin, and will def be a good basis to start from. 

In terms of custom glsl nodes, one big issue I'm having is getting access to the glsl source for some of the shaders.
Lets say I wanted to modify the diffuse or toon glsl shader (and basically write my own modified one.) I can't seem to find the locations of the source glsl nodes / glsl code.

Quickly adding the glsl into the source of a malt tree node is fine but 

1) I'm not entirely sure on how I'm supposed to write the actual setup, again I wish examples existed (or rather were up to date)
2) I'd also like to know how to write a node from scratch that points to glsl itself (rather then overwriting the source of a malt tree graph

**Reply by Kolupsy** [reply]
1. I think miguel stated that when using the NPR Pipeline it makes more sense to write custom nodes than write an entire mesh shader in glsl (please correct me if I am wrong @pragma37 ). Finding the source code for malt shaders is actually pretty easy: You just need to type the name of the function you want to inspect into the F3 search menu and follow the path shown in the Malt source code. In this example you want to find the NPR Pipeline > NPR Shading > toon_shading function

![image](https://user-images.githubusercontent.com/88381426/175523432-7f5ba441-75cc-412b-b75f-489f0529f34e.png)

2. This is what I tried to describe above. You either use the fast version of linking a graph directly to a glsl file or setting up a plugin (which you can then use in any project)

**Reply by Dai696** [reply]
Okay this is great, I'd like to take the toon shader and add ambient lighting to the shadow part. Since I can't find an easy way to add ambient lighting via the nodes. I don't know how to access world ambient light data, or light data period.

It'd be nice to be able to get light direction and feed that into the fresnel (rim light) for example. 

Also I just realized you do provide an explanation of how to write the nodes/glsl part in ur own plugin, tho I don't think I saw an example of how the glsl part works.

**Reply by Kolupsy** [reply]
> I don't know how to access world ambient light data

There is no ambient light per se in the NPR Pipeline. You can see when you follow the pipeline graph that the background gets added after all the objects in the scene have been rendered

> Also I just realized you do provide an explanation of how to write the nodes/glsl part in ur own plugin, tho I don't think I saw an example of how the glsl part works.

The explanation is specific to the plugin or provides a template of how to make a plugin of the same/similar structure. Are you already familiar with writing glsl? 

You can check out this file to see how you can accumulate the lighting of each light in the scene: https://github.com/bnpr/Malt/blob/Development/Malt/Shaders/Lighting/Lighting.glsl

**Reply by Dai696** [reply]
Ah strange, what is the "Ambient Light" value in the material for then?

Also yup, I'm familiar with glsl (tho I work with hlsl more in Unity).

**Reply by Kolupsy** [reply]
> Ah strange, what is the "Ambient Light" value in the material for then?

What are you reffering to here? I cant seem to recall any reference of ambient light in malt?

> Also yup, I'm familiar with glsl (tho I work with hlsl more in Unity).

Ok very good, luckily they are very similar. 

About that Ambient light issue: if its just a simple color then you could just add it to any lighting or shading

**Reply by Dai696** [reply]
Oops I probably mean Ambient Color.

Changing these don't seem to effect anything and I always thought it was odd to have these in the material like this.
![image](https://user-images.githubusercontent.com/17523069/175536906-f09df7d7-5033-46a2-995b-784952ace1f1.png)

> About that Ambient light issue: if its just a simple color then you could just add it to any lighting or shading

Yeah its seeming more and more that, that's what I'll have to do.

**Reply by Dai696** [reply]
Oh one last thing, I recall reading about a pbr example? Does that exist? (I'm trying to re-create the shader I use in Unity, in blender for modeling purposes. Test normals and stuff)

**Reply by Kolupsy** [reply]
You can make yourself a simple PBR shader by combining different 'base shaders' from the NPR pipeline. It gets difficult with metallic materials because the NPR Pipeline does not have screen space reflections (yet?) and - again - no default ambient lighting (which are often the key to the metallic look). 
![image](https://user-images.githubusercontent.com/88381426/175540618-8e11c5fc-cb49-480b-9948-7da1cdaf5a99.png)

**Reply by Dai696** [reply]
Why did i see a node like this?

![image](https://user-images.githubusercontent.com/17523069/175544450-4b9a46e7-bf45-4e83-9a07-520e17416590.png)

**Comment by pragma37** [comment, marked as answer]
Hi! 
There are many questions here (please, try to make one thread per question) and @Kolupsy already answered most of them, but let's do a recap.

For adding custom nodes you can either add a GLSL file as a node tree specific *Local Library*:
![imagen](https://user-images.githubusercontent.com/36610999/175560211-1eaf92d3-91b8-4696-bb27-6e7dff554d42.png)

Or as part of a plugin: https://malt3d.com/Documentation/Plugins/

In both cases, any function you add to your GLSL files will be available as a node.

For examples, you can browse the built-in functions (All function nodes are generated from a plain GLSL function):
https://github.com/bnpr/Malt/tree/Development/Malt/Shaders
https://github.com/bnpr/Malt/tree/Development/Malt/Pipelines/NPR_Pipeline/Shaders


About ambient lighting. The "realistic" way to handle it is just by adding it to the other light contributions, so basically this:
![imagen](https://user-images.githubusercontent.com/36610999/175560642-4ccf4e2d-3ac4-47fd-a344-1d47b685d53d.png)

Here's an example of a basic custom shading node "from scratch" that you can use as a starting point:
````glsl
#ifdef IS_MESH_SHADER

vec3 diffuse_example(vec3 surface_color, vec3 ambient_color)
{
    vec3 result = vec3(0);

    for (int i = 0; i < LIGHTS.lights_count; i++)
    {
        ivec4 light_group = ivec4(LIGHT_GROUP_INDEX(i));
        if(!any(equal(MATERIAL_LIGHT_GROUPS, light_group))) continue;

        Light L = LIGHTS.lights[i];
        LitSurface LS = npr_lit_surface(POSITION, NORMAL, ID.x, L, i, Settings.Receive_Shadow, Settings.Self_Shadow);

        result += max(0, LS.NoL) * LS.light_color * LS.shadow_multiply;
    }

    result += ambient_color;
    result *= surface_color;

    return result;
}

#endif
````

>If vec2 -> vec4, it'll add 2 extra 0 inputs to it
>
>If vec4->vec2, it'll ignore the last 2 values.

You can already do that, and even add your own custom conversions:
https://github.com/bnpr/Malt/discussions/182#discussioncomment-1706241

**Reply by Dai696** [reply]
Thank you, I appreciate the detailed answer! 

> You can already do that, and even add your own custom conversions:

Still doesn't fully explain why we cant simplify the nodes down to just 1 utill library that does the conversions for us.

**Reply by pragma37** [reply]
The explanation is simply that adding new features takes time.

**Reply by Dai696** [reply]
My bad, I didn't realize that was the reason. I'm down to contribute to adding features like this if the need for help is required!

**Reply by pragma37** [reply]
Sure! If you feel like contributing a templated math node that would be great.
